### PR TITLE
Validator services adjust with system time

### LINKED ISSUE #2052: VC time updates
## Description

The `DutiesService` (and others) in the VC presently doing their polling by waiting until the beginning of a slot then starting a repeating 12s timer:

https://github.com/sigp/lighthouse/blob/65dcdc361b0c91dabe53c886e32aa7952251be68/validator_client/src/duties_service.rs#L471-L478

This generally works well, however it doesn't account for the scenario where the time is incorrect and changes during operation. E.g., if your VC is 5s slow when it starts, even if you change your clock the `DutiesService` will keep using that old time.

## Steps to Resolve

Options:

- Change to a mechanism that adjusts with the system clock.
- Check against the slot clock when the services start and log an error if their out of sync (not ideal, but interesting stop-gap for the time being).

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Fixes #2052 

## Proposed Changes

Instead of running a repeated interval timer, the duties, attestation and fork services sleep for the required duration until the next slot/epoch and then run spawn their respective tasks.

### CODE REVIEW

**Comment by paulhauner** [comment]
You should be able to fix the clippy lints by merging in `unstable` :)

**Comment by paulhauner** [comment]
Just checking to make sure this is still a draft and not ready-for-review :)

**Comment by pawanjay176** [comment]
Sorry, I lost sight of this guy :sweat_smile:  Will merge unstable and do some more testing on this.

File: validator_client/src/attestation_service.rs [hunk 1]
```diff
@@ -137,19 +137,12 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
...
-            )
-        };
-
         let executor = self.context.executor.clone();
 
         let interval_fut = async move {
-            loop {
-                interval.tick().await;
+            while let Some(duration_to_next_slot) = self.slot_clock.duration_to_next_slot() {
+                // TODO(pawan): double check timing here
```
Comment (paulhauner) [inline validator_client/src/attestation_service.rs]: TODO

File: validator_client/src/attestation_service.rs [hunk 2]
```diff
@@ -137,19 +137,12 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
...
-                slot_duration,
-            )
-        };
-
         let executor = self.context.executor.clone();
 
         let interval_fut = async move {
-            loop {
-                interval.tick().await;
+            while let Some(duration_to_next_slot) = self.slot_clock.duration_to_next_slot() {
```
Comment (paulhauner) [inline validator_client/src/attestation_service.rs]: If we happen to be unable to read the system clock then this timer is going to exit and never run again. I'm not sure why that would happen, but it seems nice to be resilient against it.

I took an approach here where I just sleep for a slot if we can't read the system clock, that gives us a chance to recover if the slot clock comes good again:

https://github.com/sigp/lighthouse/blob/46920a84e857d6c086adf4b42810ac03de94ed3e/beacon_node/beacon_chain/src/state_advance_timer.rs#L99-L107

File: validator_client/src/duties_service.rs [hunk 3]
```diff
@@ -490,9 +476,11 @@ impl<T: SlotClock + 'static, E: EthSpec> DutiesService<T, E> {
         let executor = self.inner.context.executor.clone();
 
         let interval_fut = async move {
-            loop {
-                interval.tick().await;
-                self.clone().do_update(&mut block_service_tx, &spec).await;
+            while let Some(duration_to_next_slot) = self.clone().slot_clock.duration_to_next_slot()
```
Comment (paulhauner) [inline validator_client/src/duties_service.rs]: As above.

File: validator_client/src/fork_service.rs [hunk 4]
```diff
@@ -163,8 +147,10 @@ impl<T: SlotClock + 'static, E: EthSpec> ForkService<T, E> {
         let executor = context.executor.clone();
 
         let interval_fut = async move {
-            loop {
-                interval.tick().await;
+            while let Some(duration_to_next_epoch) =
```
Comment (paulhauner) [inline validator_client/src/fork_service.rs]: As above.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Just a few comments. There's also another timer here:

https://github.com/sigp/lighthouse/blob/46920a84e857d6c086adf4b42810ac03de94ed3e/validator_client/src/notifier.rs#L27

I haven't had a chance to check out the simulator with 1s slot times, but I'll check that out once these comments are addressed :)

**Comment by paulhauner** [comment]
Just double-checking that this isn't ready for review :)

File: validator_client/src/attestation_service.rs (hunk 2 above)
Comment (pawanjay176) [reply validator_client/src/attestation_service.rs]: Nice idea. Fixed this.

**Comment by pawanjay176** [comment]
Sorry forgot to mark it. It is ready for review!

**Comment by michaelsproul** [comment]
I'm really keen to get this merged @pawanjay176, it looks great, but @paulhauner just went and revamped the whole duties service in the meantime :weary:

There's a user on Discord who needs this PR, I think..

**Comment by paulhauner** [comment]
I can take care of merging this down once we get my optimise PR in :)

**Comment by paulhauner** [comment]
Closing this since I merged it into #2243. Thanks for your work @pawanjay176!

**Pull request closed without merging**